fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/mixin/entity/event/PlayerManagerMixin.java

```diff
@@ -32,4 +38,19 @@ abstract class PlayerManagerMixin {
 	private void afterRespawn(ServerPlayerEntity oldPlayer, boolean alive, CallbackInfoReturnable<ServerPlayerEntity> cir) {
 		ServerPlayerEvents.AFTER_RESPAWN.invoker().afterRespawn(oldPlayer, cir.getReturnValue(), alive);
 	}
+
+	@Inject(method = "onPlayerConnect", at = @At(value = "INVOKE", target = "net/minecraft/server/network/ServerPlayerEntity.setWorld(Lnet/minecraft/server/world/ServerWorld;)V", shift = At.Shift.AFTER))
+	private void afterWorldSet(ClientConnection connection, ServerPlayerEntity player, CallbackInfo ci) {
+		ServerPlayerEvents.BEFORE_SPAWN.invoker().beforeSpawn(player);
+	}
+
+	@ModifyVariable(method = "onPlayerConnect", name = "serverWorld2", at = @At(value = "INVOKE", target = "net/minecraft/server/network/ServerPlayerEntity.setWorld(Lnet/minecraft/server/world/ServerWorld;)V", shift = At.Shift.AFTER))
+	private ServerWorld fixServerWorld(ServerWorld world, ClientConnection connection, ServerPlayerEntity player) {
+		return player.getWorld();
+	}
+
+	@ModifyVariable(method = "onPlayerConnect", name = "registryKey", at = @At(value = "INVOKE", target = "net/minecraft/server/network/ServerPlayerEntity.setWorld(Lnet/minecraft/server/world/ServerWorld;)V", shift = At.Shift.AFTER))
```

Comment:
Same as above. Also `registryKey` isn't used after where you're injecting (`setWorld`), but I guess leaving it here is fine.